docs/proposals/20-ha-control-planes.md

```diff
@@ -0,0 +1,2624 @@
+  * Shoot with `non-multi-zonal` control plane can be scheduled either on `non-multi-zonal` seed or `multi-zonal` seed. In order to find the best fit - first check if there is any available seed in `non-multi-zonal` seeds. If there is one then that is assigned. If there is no available seed in `non-multi-zonal` seeds then look for an available seed in `multi-zonal` seeds.
+
+The above selection algorithm should be applied only after applying the [strategy](https://github.com/gardener/gardener/blob/4e2d28f2af7093eb94819652a6f4709b5fbfaf06/pkg/scheduler/controller/shoot/reconciler.go#L256-L274). This ensures that strategy based selection e.g `region affinity` or `minimum distance` is applied first and then selection of a `single-zone` or `multi-zone` seed cluster is made to optimize the geographic separation between the shoot cluster and the shoot cluster control plane.
+
+**Case #2**
+
+A shoot has a pre-defined non-HA seed. A change has been made to the shoot spec, setting control HA to `multi-zone`. Garden scheduler needs to react to the change in the HA configuration for the shoot control plane.
+
+_Proposed Changes_
+
+* [scheduler manager](https://github.com/gardener/gardener/blob/4dbba56e24b5999fd728cb84b09a4d6d54f64479/pkg/scheduler/controller/shoot/scheduler.go#L57) should remove the predicate `ShootIsUnassigned` and instead define a new predicate which allows a `reconcile.Request` to be created if one of the following is satisified:
+  * Shoot does not have any assigned seed
+  * Shoot spec has changed requesting an uprade from non-HA control plane to a HA control plane. This can be checked by checking the HA configuration in the spec and compare it with the `status` to determine if the upgrade requested is from a non-HA shoot control plane to a HA shoot control plane.
+* [reconciler](https://github.com/gardener/gardener/blob/4dbba56e24b5999fd728cb84b09a4d6d54f64479/pkg/scheduler/controller/shoot/reconciler.go#L52-L99) - function now needs to additionally check if an upgrade request has been made. If it is a valid upgrade request for the shoot control plane then the reconciler should unassign the current seed for this shoot. A new seed appropriate to the HA control plane requirement will be assigned. In this case since there is a change in the seed assignment, `control-plane-migration` flow will kick in and will commence the migration of the control plane from the non-HA seed to a HA seed.
+
+**Unbalanced usage of zones in a multi-zonal seed cluster**
+
+It is possible that `single-zone` shoot control planes are scheduled onto `multi-zonal` seed clusters. This can result in unbalanced utilization of  zones and node groups (in gardener terms this maps to `MachineDeployment`). Optimal zone selection for `single-zone` shoot control plane is currently not possible. Selection of zone and node-group will be best effort.
+
```

Comment:
See also https://github.com/gardener/gardener/pull/6102#issuecomment-1176252934, specifically:

> We have an active interest to place all 3 ETCDs of an HA shoot (not multi-zonal) in one zone to avoid networking costs. This will require some zone pinning, not only for them, but probably for the entire control plane. However, since this costs some effort to build and is generally more fragile than distributing the control plane just like any other HA multi-zonal shoot, we may decide at first to not write this cost optimisation logic. Also, because I have some doubts people will pick that option. I would assume, in most cases it's either non-HA shoot or HA multi-zonal shoot. It's only HA shoots where the end user knowingly places his nodes in a 1/2-zone region (e.g. KSA, Switzerland, ...) and there we would have 1-zone seeds anyway, so no need to optimise anything for 3-zone seeds either, I believe.

Reply:
If we don't implement zone pinning from the beginning then the `single-zone` option for shoot clusters running on a `multi-zonal` seed is no longer an option because the control-plane components can be provisioned across multiple-zones.

> It's only HA shoots where the end user knowingly places his nodes in a 1/2-zone region (e.g. KSA, Switzerland, ...) and there we would have 1-zone seeds anyway

We interpret this statement as all the `ManagedSeed`s can only be setup with odd number of zones. Do you agree?

To our understanding, the `single-zone` use-case is only supported for regions that have less than three zones. So if we expect that `single-zone` shoots on `multi-zonal` seeds won't be useful for consumers, the question is if we should drop this requirement completely. WDYT?

We see the following simplification:
- Enforcing that `ManagedSeed`s have an odd number of zones.
- Remove the option of `{single-zone, multi-zone}` for shoots and only provider `highAvailability.enabled: {true, false}`

Reply:
> We interpret this statement as all the ManagedSeeds can only be setup with odd number of zones. Do you agree?

1 or 3, yes.

Reply:
Yes, of course shoots 😅 
Thanks for the answer. Then I guess we'll stick to the plan and elaborate on the zone pinning in the revised version of the GEP.